### PR TITLE
Rewrite in-memory Merkle tree tests

### PR DESCRIPTION
**Pull request by pavelkalinnikov**
This change completes the TODO of rewriting the unit tests for the in-memory `Tree` type.
The new tests are mostly a rework of the previous tests.

The methodology of testing remains the same:
1) There is a reference implementation of root hash and proofs computation, translated from
definitions in RFC 6962. It is tested using "golden" tests and sanity checks.
2) The `Tree` type is tested exhaustively, to make sure it produces the same result as the
reference implementation.

The goal is to move the `Tree` type and tests to https://github.com/transparency-dev/merkle
`testonly` package, as it can be widely reused for tests.

### CODE REVIEW

**Comment by pavelkalinnikov** [comment]
Recommendation for reviewers: Review the latest state of this PR. The chain of commits reflects the iterative process of this refactoring, and shows that all the transitions are equivalent, but you may choose to trust me on that :)

File: internal/merkle/inmemory/reference_test.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	// subtrees are consistent. The right subtree doesn't contain the root of
+	// size1, so set haveRoot1 = false. Record the hash of the left subtree
+	// (equal in both trees).
+	return append(
+		refConsistencyProof(entries[split:], size2-split, size1-split, hasher, false),
+		refRootHash(entries[:split], hasher))
+}
+
+// downToPowerOfTwo returns the largest power of two smaller than x.
+func downToPowerOfTwo(x uint64) uint64 {
```
Comment (jiggoha) [inline internal/merkle/inmemory/reference_test.go:108]: May I suggest sqrtFloor as an alternative name?

File: internal/merkle/inmemory/tree_test.go [hunk 2]
```diff
@@ -16,634 +16,190 @@ package inmemory
...
-
-// Some paths for the reference tree.
-type pathTestVector struct {
-	leaf       uint64
-	snapshot   uint64
-	testVector []string
-}
-
-// Generated from C++ ReferenceMerklePath, not the Go one so we can verify
-// that they are both producing the same paths in a sanity test.
```
Comment (jiggoha) [inline internal/merkle/inmemory/tree_test.go:64]: Can we move this comment and the other similar ones to the `testonly` package? It would helpful there for understanding where these constants were coming from.

File: internal/merkle/inmemory/tree_test.go [hunk 3]
```diff
@@ -16,634 +16,190 @@ package inmemory
...
-// This is the hash of an empty string
-var emptyTreeHashValue = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
-
-// Inputs to the reference tree, which has eight leaves.
-var leafInputs = [][]byte{
-	hx(""), hx("00"), hx("10"), hx("2021"), hx("3031"), hx("40414243"),
-	hx("5051525354555657"), hx("606162636465666768696a6b6c6d6e6f"),
-}
-
-// Incremental roots from building the reference tree from inputs leaf-by-leaf.
```
Comment (jiggoha) [inline internal/merkle/inmemory/tree_test.go:43]: Same comment

File: internal/merkle/inmemory/tree_test.go [hunk 4]
```diff
@@ -16,634 +16,190 @@ package inmemory
...
-	},
-}
-
-type proofTestVector struct {
-	snapshot1 uint64
-	snapshot2 uint64
-	proof     []string
-}
-
-// Generated from ReferenceSnapshotConsistency in C++ version.
```
Comment (jiggoha) [inline internal/merkle/inmemory/tree_test.go:108]: Same comment

File: internal/merkle/inmemory/reference_test.go [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+// The reference Merkle tree hashing and proof algorithms in this file directly
+// implement the definitions from RFC 6962 [1]. We use this implementation only
+// for testing correctness of other more flexible and performant algorithms,
+// such as the in-memory Tree type and compact ranges.
+//
+// [1] https://datatracker.ietf.org/doc/html/rfc6962#section-2
+
+// refRootHash returns the root hash of a Merkle tree with the given entries.
+// This is a reference implementation for cross-checking.
+func refRootHash(entries [][]byte, hasher merkle.LogHasher) []byte {
```
Comment (jiggoha) [inline internal/merkle/inmemory/reference_test.go:38]: Do the functions here need to be prepended with "ref"? IIUC, "reference" here refers to the fact that this is returns an example/specific case for a root hash. However, there is nothing specific about this code, since it takes `entries` as a param. Thus, is it inaccurate to just call it `rootHash`?

File: internal/merkle/inmemory/reference_test.go (hunk 1 above)
Comment (pavelkalinnikov) [reply internal/merkle/inmemory/reference_test.go:108]: I see where your idea comes from, but it's not quite a square root, it returns the highest power of 2 which is strictly less than x.

E.g., for 9 it returns 8, whereas the square root of 9 is 3.

This name is how this func was called before the refactoring.

File: internal/merkle/inmemory/tree_test.go (hunk 2 above)
Comment (pavelkalinnikov) [reply internal/merkle/inmemory/tree_test.go:64]: My plan is:
- Complete this PR.
- Move this code to `testonly` in https://transparency-dev/merkle.
- Use these and other constants as "golden" tests there.

Looks inline with your suggestion?

File: internal/merkle/inmemory/reference_test.go (hunk 5 above)
Comment (pavelkalinnikov) [reply internal/merkle/inmemory/reference_test.go:38]: "reference" here mean that the implementation is reference (not specific values/hashes). See the comment in lines 29-34.

Comment (pavelkalinnikov) [reply internal/merkle/inmemory/reference_test.go:38]: To test various Merkle tree implementation, there are a couple of techniques:
- "golden" tests: use specific inputs and outputs to check that it matches exactly. Pros: no way to corrupt the "expected" result, cons: need to consider lots of corner cases manually.
- "reference" implementation: compare our implementation to ones that are most simple/intuitive (but maybe less performanct) and follow definitions closely; we know that these implementations probably work, so we can run an exhaustive test that verifies new algorithm on all possible inputs.

Comment (pavelkalinnikov) [reply internal/merkle/inmemory/reference_test.go:38]: So, the methodology is:
- Make reference implementations (this file).
- Test them against golden tests (this file).
- Test any other implementation exhaustively against the reference implementation (other files).
  - Note: Other tests will indirectly "inherit" all the golden tests, so there is no need to repeat the golden testing there.

Comment (pavelkalinnikov) [reply internal/merkle/inmemory/reference_test.go:38]: This way, we get Pros of both approaches, and minimise Cons.

Comment (pavelkalinnikov) [reply internal/merkle/inmemory/reference_test.go:38]: This also partly addresses you suggestion below of exposing the "known good" values. It can be done, but is not strictly necessary in the described methodology.

Comment (jiggoha) [reply internal/merkle/inmemory/reference_test.go:38]: Gotcha, thanks. Re-reading your comments in the code now makes a lot more sense to me.

File: internal/merkle/inmemory/reference_test.go (hunk 1 above)
Comment (jiggoha) [reply internal/merkle/inmemory/reference_test.go:108]: Ah, I did not look at this carefully. Fair enough.

File: internal/merkle/inmemory/tree_test.go (hunk 2 above)
Comment (jiggoha) [reply internal/merkle/inmemory/tree_test.go:64]: SGTM

**Review by jiggoha — APPROVED** [review]
